pyteal/ast/scratchvar.py

```diff
@@ -47,5 +54,86 @@ def load(self) -> ScratchLoad:
         """Load value from Scratch Space"""
         return self.slot.load(self.type)
 
+    def index(self) -> Expr:
+        return self.slot.index()
+
 
 ScratchVar.__module__ = "pyteal"
+
+
+class DynamicScratchVar(ScratchVar):
+    """
+    Example of Dynamic Scratch space whereby the slot index is picked up from the stack:
+        .. code-block:: python1
+
+            player_score = DynamicScratchVar(TealType.uint64)
+
+            wilt = ScratchVar(TealType.uint64, 129)
+            kobe = ScratchVar(TealType.uint64)
+            dt = ScratchVar(TealType.uint64, 131)
+
+            seq = Seq(
+                player_score.set_index(wilt),
+                player_score.store(Int(100)),
+                player_score.set_index(kobe),
+                player_score.store(Int(81)),
+                player_score.set_index(dt),
+                player_score.store(Int(73)),
+                Assert(player_score.load() == Int(73)),
+                Assert(player_score.index() == Int(131)),
+                player_score.set_index(wilt),
+                Assert(player_score.load() == Int(100)),
+                Assert(player_score.index() == Int(129)),
+                Int(100),
+            )
+    """
+
+    def __init__(self, ttype: TealType = TealType.anytype):
+        """Create a new DynamicScratchVar which references other ScratchVar's
+
+        Args:
+            ttype (optional): The type that this variable can hold. Defaults to TealType.anytype.
+        """
+        self.type = ttype
+        self._indexer = ScratchVar(TealType.uint64)
+        self.slot = self._indexer.slot
+
+    def set_index(self, index_var: ScratchVar) -> Expr:
+        """Set this DynamicScratchVar to reference the provided `indexVar`.
+        Followup `store`, `load` and `index` operations will use the provided `indexVar` until
```

Comment:
@tzaffi Minor:  I think there's a mismatch when referring to the method parameters.

```suggestion
        """Set this DynamicScratchVar to reference the provided `index_var`.
        Followup `store`, `load` and `index` operations will use the provided `index_var` until
```